File: src/frontend/nfe/lib/app/data/model/nfe_cana_deducoes_safra_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:nfe/app/data/model/model_imports.dart';

class NfeCanaDeducoesSafraModel {
	int? id;
	int? idNfeCana;
	String? decricao;
	double? valorDeducao;
	double? valorFornecimento;
	double? valorTotalDeducao;
	double? valorLiquidoFornecimento;
	NfeCanaModel? nfeCanaModel;

	NfeCanaDeducoesSafraModel({
		this.id,
		this.idNfeCana,
		this.decricao,
		this.valorDeducao,
		this.valorFornecimento,
		this.valorTotalDeducao,
		this.valorLiquidoFornecimento,
		this.nfeCanaModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'decricao',
		'valor_deducao',
		'valor_fornecimento',
		'valor_total_deducao',
		'valor_liquido_fornecimento',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Decricao',
		'Valor Deducao',
		'Valor Fornecimento',
		'Valor Total Deducao',
		'Valor Liquido Fornecimento',
	];

	NfeCanaDeducoesSafraModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idNfeCana = jsonData['idNfeCana'];
		decricao = jsonData['decricao'];
		valorDeducao = jsonData['valorDeducao']?.toDouble();
		valorFornecimento = jsonData['valorFornecimento']?.toDouble();
		valorTotalDeducao = jsonData['valorTotalDeducao']?.toDouble();
		valorLiquidoFornecimento = jsonData['valorLiquidoFornecimento']?.toDouble();
		nfeCanaModel = jsonData['nfeCanaModel'] == null ? NfeCanaModel() : NfeCanaModel.fromJson(jsonData['nfeCanaModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idNfeCana'] = idNfeCana != 0 ? idNfeCana : null;
		jsonData['decricao'] = decricao;
		jsonData['valorDeducao'] = valorDeducao;
		jsonData['valorFornecimento'] = valorFornecimento;
		jsonData['valorTotalDeducao'] = valorTotalDeducao;
		jsonData['valorLiquidoFornecimento'] = valorLiquidoFornecimento;
		jsonData['nfeCanaModel'] = nfeCanaModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idNfeCana = plutoRow.cells['idNfeCana']?.value;
		decricao = plutoRow.cells['decricao']?.value;
		valorDeducao = plutoRow.cells['valorDeducao']?.value?.toDouble();
		valorFornecimento = plutoRow.cells['valorFornecimento']?.value?.toDouble();
		valorTotalDeducao = plutoRow.cells['valorTotalDeducao']?.value?.toDouble();
		valorLiquidoFornecimento = plutoRow.cells['valorLiquidoFornecimento']?.value?.toDouble();
		nfeCanaModel = NfeCanaModel();
		nfeCanaModel?.safra = plutoRow.cells['nfeCanaModel']?.value;
	}	

	NfeCanaDeducoesSafraModel clone() {
		return NfeCanaDeducoesSafraModel(
			id: id,
			idNfeCana: idNfeCana,
			decricao: decricao,
			valorDeducao: valorDeducao,
			valorFornecimento: valorFornecimento,
			valorTotalDeducao: valorTotalDeducao,
			valorLiquidoFornecimento: valorLiquidoFornecimento,
		);			
	}

	
}

File: src/backend/nfe/src/model/nfe_cana_deducoes_safra_model.py
from src import db
from src.model.nfe_cana_model import NfeCanaModel


class NfeCanaDeducoesSafraModel(db.Model):
    __tablename__ = 'nfe_cana_deducoes_safra'

    id = db.Column(db.Integer, primary_key=True)
    decricao = db.Column(db.String(60))
    valor_deducao = db.Column(db.Float)
    valor_fornecimento = db.Column(db.Float)
    valor_total_deducao = db.Column(db.Float)
    valor_liquido_fornecimento = db.Column(db.Float)
    id_nfe_cana = db.Column(db.Integer, db.ForeignKey('nfe_cana.id'))

    nfe_cana_model = db.relationship('NfeCanaModel', foreign_keys=[id_nfe_cana])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_nfe_cana = data.get('idNfeCana')
        self.decricao = data.get('decricao')
        self.valor_deducao = data.get('valorDeducao')
        self.valor_fornecimento = data.get('valorFornecimento')
        self.valor_total_deducao = data.get('valorTotalDeducao')
        self.valor_liquido_fornecimento = data.get('valorLiquidoFornecimento')

    def serialize(self):
        return {
            'id': self.id,
            'idNfeCana': self.id_nfe_cana,
            'decricao': self.decricao,
            'valorDeducao': self.valor_deducao,
            'valorFornecimento': self.valor_fornecimento,
            'valorTotalDeducao': self.valor_total_deducao,
            'valorLiquidoFornecimento': self.valor_liquido_fornecimento,
            'nfeCanaModel': self.nfe_cana_model.serialize() if self.nfe_cana_model else None,
        }